File: src/api/decoders/teaching_operations.ts
// Strict browser decoders for generated Course teaching-operation DTOs.

import { MAX_ASSESSMENT_ATTEMPT_LIMIT } from "../../../generated/api/MAX_ASSESSMENT_ATTEMPT_LIMIT";
import { MAX_ASSESSMENT_ATTEMPT_TIME_LIMIT_SECONDS } from "../../../generated/api/MAX_ASSESSMENT_ATTEMPT_TIME_LIMIT_SECONDS";
import { MAX_TEACHING_DISPLAY_LABEL_UNICODE_SCALARS } from "../../../generated/api/MAX_TEACHING_DISPLAY_LABEL_UNICODE_SCALARS";
import { MAX_TEACHING_PAGE_SIZE } from "../../../generated/api/MAX_TEACHING_PAGE_SIZE";
import type { HypotheticalStudentViewScenarioModifiers } from "../../../generated/api/HypotheticalStudentViewScenarioModifiers";
import type { CourseInvitationTerminalActionRequest } from "../../../generated/api/CourseInvitationTerminalActionRequest";
import type { AccountTimeZone } from "../../../generated/api/AccountTimeZone";
import type { PendingCourseInvitationsPage } from "../../../generated/api/PendingCourseInvitationsPage";
import {
  DecodeError,
  decodeNullable,
  decodeRecord,
  decodeSafeInteger,
  decodeString,
  decodeStringEnum,
  decodeUuid,
} from "../decoder";
import {
  decodeBoundedArray,
  decodeCursor,
  decodeTimestamp,
  field,
  requireOnlyFields,
} from "./shared";

const INVITATION_STATES = ["pending", "expired", "accepted", "declined", "revoked"] as const;

function closed(
  value: unknown,
  path: string,
  keys: ReadonlyArray<string>,
): Record<string, unknown> {
  const record = decodeRecord(value, path);
  requireOnlyFields(record, path, keys);
  for (const key of keys) field(record, key, path);
  return record;
}

function boundedTrimmedText(value: unknown, path: string, maximum: number): string {
  const text = decodeString(value, path);
  if (text.trim() !== text || text.trim().length === 0 || Array.from(text).length > maximum) {
    throw new DecodeError(path, `trimmed nonblank text no longer than ${maximum} Unicode scalars`);
  }
  return text;
}

function canonicalPositivePostgresBigint(value: unknown, path: string): string {
  const parsed = decodeString(value, path);
  if (!/^[1-9][0-9]{0,18}$/u.test(parsed) || BigInt(parsed) > 9_223_372_036_854_775_807n) {
    throw new DecodeError(path, "a canonical positive PostgreSQL bigint decimal");
  }
  return parsed;
}

function pageCursor(value: unknown, path: string): string | null {
  return decodeNullable(value, path, decodeCursor);
}

function displayTimeZone(value: unknown, path: string): AccountTimeZone {
  const timeZone = decodeString(value, path);
  if (timeZone.length === 0 || timeZone.length > 255 || timeZone.trim() !== timeZone) {
    throw new DecodeError(path, "a bounded exact IANA time-zone name");
  }
  try {
    new Intl.DateTimeFormat(undefined, { timeZone });
  } catch {
    throw new DecodeError(path, "a browser-supported IANA time-zone name");
  }
  return timeZone;
}

function positiveInteger(value: unknown, path: string, maximum: number): number {
  const parsed = decodeSafeInteger(value, path);
  if (parsed < 1 || parsed > maximum) {
    throw new DecodeError(path, `a positive integer no larger than ${maximum}`);
  }
  return parsed;
}

function courseLocalDateAndTime(value: unknown, path: string): string {
  const text = decodeString(value, path);
  const match = /^(\d{4})-(\d{2})-(\d{2})T(\d{2}):(\d{2}):(\d{2})\.(\d{3})$/u.exec(text);
  if (match === null)
    throw new DecodeError(path, "an exact valid YYYY-MM-DDTHH:MM:SS.sss local date-time");
  const year = Number(match[1]);
  const month = Number(match[2]);
  const day = Number(match[3]);
  const hour = Number(match[4]);
  const minute = Number(match[5]);
  const second = Number(match[6]);
  const leap = year % 4 === 0 && (year % 100 !== 0 || year % 400 === 0);
  const monthLengths = [31, leap ? 29 : 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31];
  const monthLength = monthLengths[month - 1];
  if (
    year === 0 ||
    monthLength === undefined ||
    day < 1 ||
    day > monthLength ||
    hour > 23 ||
    minute > 59 ||
    second > 59
  ) {
    throw new DecodeError(path, "an exact valid YYYY-MM-DDTHH:MM:SS.sss local date-time");
  }
  return text;
}

function timePatch(
  value: unknown,
  path: string,
): HypotheticalStudentViewScenarioModifiers["adjustment"]["available_at"] {
  const record = decodeRecord(value, path);
  const kind = decodeString(field(record, "kind", path), `${path}.kind`);
  switch (kind) {
    case "inherit":
    case "unrestricted":
      requireOnlyFields(record, path, ["kind"]);
      return { kind };
    case "set":
      requireOnlyFields(record, path, ["kind", "value"]);
      return { kind, value: courseLocalDateAndTime(field(record, "value", path), `${path}.value`) };
    default:
      throw new DecodeError(`${path}.kind`, "one of inherit, set, unrestricted");
  }
}

function limitPatch(
  value: unknown,
  path: string,
  maximum: number,
): HypotheticalStudentViewScenarioModifiers["adjustment"]["assessment_attempt_time_limit_seconds"] {
  const record = decodeRecord(value, path);
  const kind = decodeString(field(record, "kind", path), `${path}.kind`);
  switch (kind) {
    case "inherit":
    case "unrestricted":
      requireOnlyFields(record, path, ["kind"]);
      return { kind };
    case "set":
      requireOnlyFields(record, path, ["kind", "value"]);
      return {
        kind,
        value: positiveInteger(field(record, "value", path), `${path}.value`, maximum),
      };
    default:
      throw new DecodeError(`${path}.kind`, "one of inherit, set, unrestricted");
  }
}

function accommodationAdjustment(
  value: unknown,
  path: string,
): HypotheticalStudentViewScenarioModifiers["adjustment"] {
  const record = closed(value, path, [
    "available_at",
    "due_at",
    "closes_at",
    "assessment_attempt_time_limit_seconds",
    "attempt_limit",
  ]);
  return {
    available_at: timePatch(record.available_at, `${path}.available_at`),
    due_at: timePatch(record.due_at, `${path}.due_at`),
    closes_at: timePatch(record.closes_at, `${path}.closes_at`),
    assessment_attempt_time_limit_seconds: limitPatch(
      record.assessment_attempt_time_limit_seconds,
      `${path}.assessment_attempt_time_limit_seconds`,
      MAX_ASSESSMENT_ATTEMPT_TIME_LIMIT_SECONDS,
    ),
    attempt_limit: limitPatch(
      record.attempt_limit,
      `${path}.attempt_limit`,
      MAX_ASSESSMENT_ATTEMPT_LIMIT,
    ),
  };
}

function decodeAccommodationAdjustmentWrite(
  value: unknown,
  path: string,
): HypotheticalStudentViewScenarioModifiers {
  const record = closed(value, path, ["mode", "adjustment"]);
  return {
    mode: decodeStringEnum(record.mode, `${path}.mode`, ["extend_only", "replace"] as const),
    adjustment: accommodationAdjustment(record.adjustment, `${path}.adjustment`),
  };
}

export function decodeHypotheticalStudentViewScenarioModifiers(
  value: unknown,
  path = "request",
): HypotheticalStudentViewScenarioModifiers {
  return decodeAccommodationAdjustmentWrite(value, path);
}

export function decodePendingCourseInvitationsPage(
  value: unknown,
  path = "response",
): PendingCourseInvitationsPage {
  const record = closed(value, path, ["displayTimeZone", "invitations", "nextCursor"]);
  return {
    // ASVS 8.2.3/14.2.6: viewer-owned metadata belongs to the outer page only.
    displayTimeZone: displayTimeZone(record.displayTimeZone, `${path}.displayTimeZone`),
    invitations: decodeBoundedArray(
      record.invitations,
      `${path}.invitations`,
      MAX_TEACHING_PAGE_SIZE,
      (entry, entryPath) => {
        const invitation = closed(entry, entryPath, [
          "id",
          "courseLabel",
          "state",
          "expiresAt",
          "state_precondition",
        ]);
        return {
          id: decodeUuid(invitation.id, `${entryPath}.id`),
          courseLabel: boundedTrimmedText(
            invitation.courseLabel,
            `${entryPath}.courseLabel`,
            MAX_TEACHING_DISPLAY_LABEL_UNICODE_SCALARS,
          ),
          state: decodeStringEnum(invitation.state, `${entryPath}.state`, INVITATION_STATES),
          expiresAt: decodeTimestamp(invitation.expiresAt, `${entryPath}.expiresAt`),
          state_precondition: canonicalPositivePostgresBigint(
            invitation.state_precondition,
            `${entryPath}.state_precondition`,
          ),
        };
      },
    ),
    nextCursor: pageCursor(record.nextCursor, `${path}.nextCursor`),
  };
}

export function decodeCourseInvitationTerminalActionRequest(
  value: unknown,
  path = "request",
): CourseInvitationTerminalActionRequest {
  const record = closed(value, path, ["action"]);
  return {
    action: decodeStringEnum(record.action, `${path}.action`, ["accept", "decline"] as const),
  };
}

File: crates/question_model/src/teaching_operations.rs
//! Strict browser/server contracts for teaching operations.
//!
//! The types here use only human route references and display labels.  They
//! deliberately exclude external-affiliation IDs, UUIDs, email, policy inputs, jobs, object
//! keys, recipient lists, Answer Key facts, and clock authority. A server maps
//! its authorized Store/domain result into these values after resolving Active Student
//! Course Membership and Effective Assessment Policy.

use std::num::{NonZeroU32, NonZeroU64};
use std::str::FromStr;

use serde::{Deserialize, Serialize};

use crate::{
    CourseInvitationId, LocalDateAndTime, MAX_ASSESSMENT_ATTEMPT_LIMIT,
    MAX_ASSESSMENT_ATTEMPT_TIME_LIMIT_SECONDS, Timestamp,
};

pub use crate::preview_plane::HypotheticalStudentViewScenarioModifiers;

/// Maximum Unicode scalar count for an authorized account or membership label.
pub const MAX_TEACHING_DISPLAY_LABEL_UNICODE_SCALARS: usize = 200;
/// Maximum rows in one browser teaching-operations page.
pub const MAX_TEACHING_PAGE_SIZE: u32 = 100;

fn canonical_positive_postgres_bigint(value: &str) -> Result<NonZeroU64, &'static str> {
    if value.is_empty()
        || (value.len() > 1 && value.starts_with('0'))
        || !value.bytes().all(|byte| byte.is_ascii_digit())
    {
        return Err("must be a canonical positive decimal string");
    }
    value
        .parse::<u64>()
        .ok()
        .filter(|value| *value <= i64::MAX as u64)
        .and_then(NonZeroU64::new)
        .ok_or("must fit a positive PostgreSQL bigint")
}

/// One Course Invitation's current lifecycle-state precondition, represented as canonical decimal JSON.
#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize)]
#[serde(try_from = "String", into = "String")]
pub struct CourseInvitationStatePrecondition(NonZeroU64);

impl CourseInvitationStatePrecondition {
    /// Creates a positive lifecycle-state precondition that fits PostgreSQL `BIGINT`.
    pub fn new(value: u64) -> Option<Self> {
        (value <= i64::MAX as u64)
            .then(|| NonZeroU64::new(value))
            .flatten()
            .map(Self)
    }

    /// Returns the exact lifecycle-state precondition used in a strong conditional request.
    pub fn value(self) -> u64 {
        self.0.get()
    }
}

impl std::fmt::Display for CourseInvitationStatePrecondition {
    fn fmt(&self, formatter: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(formatter, "{}", self.value())
    }
}

impl FromStr for CourseInvitationStatePrecondition {
    type Err = &'static str;

    fn from_str(value: &str) -> Result<Self, Self::Err> {
        canonical_positive_postgres_bigint(value).map(Self)
    }
}

impl TryFrom<String> for CourseInvitationStatePrecondition {
    type Error = &'static str;

    fn try_from(value: String) -> Result<Self, Self::Error> {
        value.parse()
    }
}

impl From<CourseInvitationStatePrecondition> for String {
    fn from(value: CourseInvitationStatePrecondition) -> Self {
        value.to_string()
    }
}

/// One Course Roster's current change number, represented as canonical decimal JSON.
#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize)]
#[serde(try_from = "String", into = "String")]
pub struct CourseRosterChangeNumber(NonZeroU64);

impl CourseRosterChangeNumber {
    /// Creates a positive change number that fits PostgreSQL `BIGINT`.
    pub fn new(value: u64) -> Option<Self> {
        (value <= i64::MAX as u64)
            .then(|| NonZeroU64::new(value))
            .flatten()
            .map(Self)
    }

    /// Returns the exact change number used in a strong conditional request.
    pub fn value(self) -> u64 {
        self.0.get()
    }
}

impl std::fmt::Display for CourseRosterChangeNumber {
    fn fmt(&self, formatter: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(formatter, "{}", self.value())
    }
}

impl FromStr for CourseRosterChangeNumber {
    type Err = &'static str;

    fn from_str(value: &str) -> Result<Self, Self::Err> {
        canonical_positive_postgres_bigint(value).map(Self)
    }
}

impl TryFrom<String> for CourseRosterChangeNumber {
    type Error = &'static str;

    fn try_from(value: String) -> Result<Self, Self::Error> {
        value.parse()
    }
}

impl From<CourseRosterChangeNumber> for String {
    fn from(value: CourseRosterChangeNumber) -> Self {
        value.to_string()
    }
}

/// Validated, nonblank browser display label with no email semantics.
#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize)]
#[serde(try_from = "String", into = "String")]
pub struct TeachingDisplayLabel(String);

impl TeachingDisplayLabel {
    /// Returns the safe human label.
    pub fn as_str(&self) -> &str {
        &self.0
    }
}

impl TryFrom<String> for TeachingDisplayLabel {
    type Error = &'static str;

    fn try_from(value: String) -> Result<Self, Self::Error> {
        if value.trim().is_empty()
            || value.trim() != value
            || value.chars().count() > MAX_TEACHING_DISPLAY_LABEL_UNICODE_SCALARS
        {
            return Err("display label must be trimmed, nonblank, and at most 200 characters");
        }
        Ok(Self(value))
    }
}

impl From<TeachingDisplayLabel> for String {
    fn from(value: TeachingDisplayLabel) -> Self {
        value.0
    }
}

/// Closed modification behavior.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum AccommodationApplicationRuleView {
    ExtendOnly,
    Replace,
}

/// Explicit adjustment state for a resolved time field; omitted fields never mean inherit.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(tag = "kind", rename_all = "snake_case", deny_unknown_fields)]
pub enum TeachingTimeFieldPatch {
    Inherit,
    Set { value: LocalDateAndTime },
    Unrestricted,
}

/// Explicit adjustment state for a resolved positive integer field.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(tag = "kind", rename_all = "snake_case", deny_unknown_fields)]
pub enum TeachingAssessmentAttemptTimeLimitFieldPatch {
    Inherit,
    Set {
        value: TeachingAssessmentAttemptTimeLimitSeconds,
    },
    Unrestricted,
}

/// Explicit adjustment state for an attempt limit.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(tag = "kind", rename_all = "snake_case", deny_unknown_fields)]
pub enum TeachingAttemptLimitFieldPatch {
    Inherit,
    Set { value: TeachingAttemptLimit },
    Unrestricted,
}

/// Positive time limit that fits the PostgreSQL `INTEGER` policy column.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(try_from = "u32", into = "u32")]
pub struct TeachingAssessmentAttemptTimeLimitSeconds(NonZeroU32);

impl TryFrom<u32> for TeachingAssessmentAttemptTimeLimitSeconds {
    type Error = &'static str;

    fn try_from(value: u32) -> Result<Self, Self::Error> {
        NonZeroU32::new(value)
            .filter(|limit| limit.get() <= MAX_ASSESSMENT_ATTEMPT_TIME_LIMIT_SECONDS)
            .map(Self)
            .ok_or("time limit must fit the assessment policy bounds")
    }
}

impl From<TeachingAssessmentAttemptTimeLimitSeconds> for u32 {
    fn from(value: TeachingAssessmentAttemptTimeLimitSeconds) -> Self {
        value.0.get()
    }
}

/// Positive attempt limit that fits the PostgreSQL `INTEGER` policy column.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(try_from = "u32", into = "u32")]
pub struct TeachingAttemptLimit(NonZeroU32);

impl TryFrom<u32> for TeachingAttemptLimit {
    type Error = &'static str;

    fn try_from(value: u32) -> Result<Self, Self::Error> {
        NonZeroU32::new(value)
            .filter(|limit| limit.get() <= MAX_ASSESSMENT_ATTEMPT_LIMIT)
            .map(Self)
            .ok_or("attempt limit must fit the assessment policy bounds")
    }
}

impl From<TeachingAttemptLimit> for u32 {
    fn from(value: TeachingAttemptLimit) -> Self {
        value.0.get()
    }
}

/// Complete adjustment replacement: every adjustment state is explicit and closed.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case", deny_unknown_fields)]
pub struct AccommodationAdjustmentView {
    pub available_at: TeachingTimeFieldPatch,
    pub due_at: TeachingTimeFieldPatch,
    pub closes_at: TeachingTimeFieldPatch,
    pub assessment_attempt_time_limit_seconds: TeachingAssessmentAttemptTimeLimitFieldPatch,
    pub attempt_limit: TeachingAttemptLimitFieldPatch,
}

/// Pending account-owned invitation row. It intentionally contains no email.
#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct PendingCourseInvitationView {
    pub id: CourseInvitationId,
    pub course_label: TeachingDisplayLabel,
    pub state: CourseInvitationStateView,
    pub expires_at: Timestamp,
    /// Exact lifecycle-state precondition required by accept or decline `If-Match`.
    #[serde(rename = "state_precondition")]
    pub state_precondition: CourseInvitationStatePrecondition,
}

/// Closed pending-invitation lifecycle state.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize)]
#[serde(rename_all = "camelCase")]
pub enum CourseInvitationStateView {
    Pending,
    Expired,
    Accepted,
    Declined,
    Revoked,
}

/// Bounded pending invitation page.
#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct PendingCourseInvitationsPage {
    /// Authenticated Instructor's own IANA zone for rendering this page's instants.
    pub display_time_zone: crate::AccountTimeZone,
    pub invitations: Vec<PendingCourseInvitationView>,
    pub next_cursor: Option<String>,
}

/// Closed terminal action for the authenticated invitation target.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum CourseInvitationTerminalAction {
    Accept,
    Decline,
}

/// Strict terminal pending-invitation action.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct CourseInvitationTerminalActionRequest {
    pub action: CourseInvitationTerminalAction,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn state_preconditions_and_labels_are_canonical_and_bounded() {
        assert!("01".parse::<CourseInvitationStatePrecondition>().is_err());
        assert!("0".parse::<CourseInvitationStatePrecondition>().is_err());
        assert_eq!(
            "42".parse::<CourseInvitationStatePrecondition>()
                .unwrap()
                .value(),
            42
        );
        assert!("01".parse::<CourseRosterChangeNumber>().is_err());
        assert!("0".parse::<CourseRosterChangeNumber>().is_err());
        assert_eq!(
            "42".parse::<CourseRosterChangeNumber>().unwrap().value(),
            42
        );
        assert!(TeachingDisplayLabel::try_from(" ".to_owned()).is_err());
    }

    #[test]
    fn mutation_bodies_are_value_only_and_state_preconditions_are_absent() {
        let action = serde_json::to_value(CourseInvitationTerminalActionRequest {
            action: CourseInvitationTerminalAction::Accept,
        })
        .unwrap();
        assert_eq!(action, serde_json::json!({"action":"accept"}));
    }

    #[test]
    fn validated_deserialization_rejects_bounds_and_duplicate_members() {
        assert!(
            serde_json::from_str::<AccommodationAdjustmentView>(concat!(
                r#"{"available_at":{"kind":"inherit"},"due_at":{"kind":"inherit"},"#,
                r#""closes_at":{"kind":"inherit"},"assessment_attempt_time_limit_seconds":{"kind":"set","#,
                r#""value":2147483648},"attempt_limit":{"kind":"inherit"}}"#
            ))
            .is_err()
        );
    }

    #[test]
    fn pending_invitation_serializes_server_owned_expiry_without_inviter() {
        let row = PendingCourseInvitationView {
            id: "00000000-0000-0000-0000-000000000004".parse().unwrap(),
            course_label: TeachingDisplayLabel::try_from("Biochemistry".to_owned()).unwrap(),
            state: CourseInvitationStateView::Pending,
            expires_at: Timestamp::from_unix_millis(2_592_000_000),
            state_precondition: "3".parse().unwrap(),
        };
        let value = serde_json::to_value(row).unwrap();
        assert_eq!(value["expiresAt"], 2_592_000_000_i64);
        assert!(value.get("invitedBy").is_none());
    }

    #[test]
    fn invitation_pages_keep_the_viewer_zone_at_the_page_envelope() {
        let display_time_zone: crate::AccountTimeZone = "America/New_York".parse().unwrap();
        let pending_page = PendingCourseInvitationsPage {
            display_time_zone,
            invitations: vec![],
            next_cursor: None,
        };

        let page = serde_json::to_value(pending_page).unwrap();
        assert_eq!(page["displayTimeZone"], "America/New_York");
        assert_eq!(page["invitations"], serde_json::json!([]));
    }
}
